tiledb/sm/filesystem/mem_filesystem.cc

```diff
@@ -0,0 +1,423 @@
+      if (data_ == nullptr) {
+        return LOG_STATUS(Status::MemFSError(
+            string("Out of memory, cannot write to file: " + name_)));
+      }
+      memcpy(data_, data, nbytes);
+      size_ = nbytes;
+    } else {
+      /* if the file already had some data */
+      void* new_data;
+      new_data = realloc(data_, size_ + nbytes);
+
+      if (new_data == nullptr) {
+        return LOG_STATUS(Status::MemFSError(string(
+            "Out of memory, cannot append new data to file: " + name_ +
+            "with size " + "size_")));
+      }
+
+      memcpy((char*)new_data + size_, data, nbytes);
+      data_ = new_data;
+      size_ += nbytes;
+    }
+
+    return Status::Ok();
+  }
+
+  /* Outputs in a buffer nbytes of this file starting at offset */
+  Status read(
+      const uint64_t offset, void* buffer, const uint64_t nbytes) const {
+    assert(buffer);
+    unique_lock<mutex> node_lock(node_mutex_);
+
+    if (offset + nbytes > size_)
+      return LOG_STATUS(
+          Status::MemFSError("Cannot read from file; Read exceeds file size"));
+
+    memcpy(buffer, (char*)data_ + offset, nbytes);
+    return Status::Ok();
+  }
+
+ private:
+  /* the data stored in this file */
+  void* data_;
+  /* the size in bytes of the data in this file */
+  uint64_t size_;
+};
+
+class MemFilesystem::Directory : public MemFilesystem::FSNode {
+ public:
+  Directory(const string& name)
+      : FSNode(name){};
+
+  ~Directory(){};
+
+  bool is_dir() const override {
+    return true;
+  }
+
+  vector<string> ls(const string& full_path) const override {
+    vector<string> names;
+    unique_lock<mutex> node_lock(node_mutex_);
+
+    for (const auto& child : children_) {
+      names.emplace_back(full_path + child.first);
+    }
+
+    sort(names.begin(), names.end());
+    return names;
+  }
+
+  bool has_child(const string& child) const override {
+    unique_lock<mutex> node_lock(node_mutex_);
+    return (children_.count(child) != 0);
+  }
+};
+
+MemFilesystem::MemFilesystem()
+    : root_(unique_ptr<Directory>(new Directory(""))){};
+
+MemFilesystem::~MemFilesystem(){};
+
+Status MemFilesystem::create_dir(const string& path) const {
+  assert(root_);
+
+  vector<string> tokens = tokenize(path);
+  FSNode* cur = root_.get();
+  for (auto& token : tokens) {
+    if (!cur->has_child(token)) {
+      cur->children_[token] = unique_ptr<Directory>(new Directory(token));
+    } else if (!cur->is_dir()) {
+      return LOG_STATUS(Status::MemFSError(string(
+          "Cannot create directory, a file with that name exists already: " +
+          path)));
+    }
+    cur = cur->children_[token].get();
+  }
+
+  return Status::Ok();
+}
+
+Status MemFilesystem::file_size(const std::string& path, uint64_t* size) const {
+  FSNode* cur = lookup_node(path);
+  if (cur == nullptr || cur->is_dir()) {
+    return LOG_STATUS(
+        Status::MemFSError(string("Cannot get file size of :" + path)));
+  }
+
+  *size = ((File*)cur)->get_size();
```

Comment:
Let's avoid a parent-class cast and put `get_size` in the parent. We could have it return a `Status` and have the directory implementation return a non-ok Status.